src/protocol/prf_sharding/mod.rs

```diff
@@ -0,0 +1,704 @@
+            Self::ComputeDifferenceToCap(i) => COMPUTE_DIFFERENCE_TO_CAP[*i],
+            Self::ComputedCappedAttributedTriggerValueNotSaturatedCase(i) => {
+                COMPUTE_CAPPED_ATTRIBUTED_TRIGGER_VALUE_NOT_SATURATED_CASE[*i]
+            }
+            Self::ComputedCappedAttributedTriggerValueJustSaturatedCase(i) => {
+                COMPUTE_CAPPED_ATTRIBUTED_TRIGGER_VALUE_JUST_SATURATED_CASE[*i]
+            }
+        }
+    }
+}
+
+/// Sub-protocol of the PRF-sharded IPA Protocol
+///
+/// After the computation of the per-user PRF, addition of dummy records and shuffling,
+/// the PRF column can be revealed. After that, all of the records corresponding to a single
+/// device can be processed together.
+///
+/// This circuit expects to receive records from multiple users,
+/// but with all of the records from a given user adjacent to one another, and in time order.
+///
+/// This circuit will compute attribution, and per-user capping.
+///
+/// The output of this circuit is the input to the next stage: Aggregation.
+///
+/// # Errors
+/// Propagates errors from multiplications
+/// # Panics
+/// Propagates errors from multiplications
+pub async fn attribution_and_capping<C, BK, TV>(
+    sh_ctx: C,
+    input_rows: &[PrfShardedIpaInputRow<BK, TV>],
+    num_breakdown_key_bits: usize,
+    num_trigger_value_bits: usize,
+    num_saturating_sum_bits: usize,
+) -> Result<Vec<CappedAttributionOutputs>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    BK: GaloisField,
+    TV: GaloisField,
+{
+    assert!(num_saturating_sum_bits > num_trigger_value_bits);
+    assert!(num_trigger_value_bits > 0);
+    assert!(num_breakdown_key_bits > 0);
+
+    let binary_validator = sh_ctx.narrow(&Step::BinaryValidator).validator::<Gf2>();
+    // TODO: fix num total records to be not a hard-coded constant, but variable per step
+    // based on the histogram of how many users have how many records a piece
+    let binary_m_ctx = binary_validator.context().set_total_records(1);
+
+    let mut output = vec![];
+
+    assert!(!input_rows.is_empty());
+    let first_row = &input_rows[0];
+    let mut prev_prf = first_row.prf_of_match_key;
+    let mut prev_row_inputs = initialize_new_device_attribution_variables(
+        Replicated::share_known_value(&binary_m_ctx, Gf2::ONE),
+        first_row,
+        num_breakdown_key_bits,
+        num_trigger_value_bits,
+        num_saturating_sum_bits,
+    );
+    let mut i: usize = 1;
+    let mut num_users_encountered = 0;
+    let mut row_for_user = 0;
+    while i < input_rows.len() {
+        let cur_row = &input_rows[i];
+        if prev_prf == cur_row.prf_of_match_key {
+            // Do some actual computation
+            let (inputs_required_for_next_row, capped_attribution_outputs) =
+                compute_row_with_previous(
+                    binary_m_ctx.clone(),
+                    RecordId(num_users_encountered),
```

Comment:
Maybe I'm missing something about the structure, but I would have expected `RecordId(i)`.

Also, it may be necessary to arrange for `compute_row_with_previous` to be called for every row, even the first row for each match key, to satisfy the restrictions on contiguous record IDs for exchanged data.

Reply:
I'm in the process of refactoring this and all of this will change to be better